packages/backend/src/nest/storage/UserProfileStore.ts

```diff
@@ -0,0 +1,261 @@
+ * https://developer.mozilla.org/en-US/docs/Web/API/FileReader/readAsDataURL
+ * and converts it to a Uint8Array.
+ */
+export const base64DataURLToByteArray = (contents: string): Uint8Array => {
+  const [header, base64Data] = contents.split(',')
+  if (!header.startsWith('data:') || !header.endsWith(';base64')) {
+    throw new Error('Expected base64 data URI')
+  }
+  const chars = atob(base64Data)
+  const bytes = new Array(chars.length)
+  for (let i = 0; i < chars.length; i++) {
+    bytes[i] = chars.charCodeAt(i)
+  }
+  return new Uint8Array(bytes)
+}
+
+export class UserProfileStore {
+  public orbitDb: OrbitDB
+  public store: KeyValueStore<UserProfile>
+
+  // Copying OrbitDB by using dag-cbor/sha256 for converting the
+  // profile to a byte array for signing:
+  // https://github.com/orbitdb/orbitdb/blob/3eee148510110a7b698036488c70c5c78f868cd9/src/oplog/entry.js#L75-L76
+  // I think any encoding would work here.
+  public static readonly codec = dagCbor
+  public static readonly hasher = sha256
+
+  constructor(orbitDb: OrbitDB) {
+    this.orbitDb = orbitDb
+  }
+
+  public async init(emitter: EventEmitter) {
+    logger('Initializing user profiles key/value store')
+
+    this.store = await this.orbitDb.keyvalue<UserProfile>('user-profiles', {
+      replicate: false,
+      Index: UserProfileKeyValueIndex,
+      accessController: {
+        write: ['*'],
+      },
+    })
+
+    this.store.events.on('write', (_address, entry) => {
+      logger('Saved user profile locally')
+      emitter.emit(StorageEvents.LOADED_USER_PROFILES, {
+        profiles: [entry.payload.value],
+      })
+    })
+
+    this.store.events.on('ready', async () => {
+      logger('Loaded user profiles to memory')
+      emitter.emit(StorageEvents.LOADED_USER_PROFILES, {
+        profiles: await this.getUserProfiles(),
+      })
+    })
+
+    this.store.events.on('replicated', async () => {
+      logger('Replicated user profiles')
+      emitter.emit(StorageEvents.LOADED_USER_PROFILES, {
+        profiles: await this.getUserProfiles(),
+      })
+    })
+
+    // @ts-expect-error - OrbitDB's type declaration of `load` lacks 'options'
+    await this.store.load({ fetchEntryTimeout: 15000 })
+  }
+
+  public getAddress() {
+    return this.store?.address
+  }
+
+  public async close() {
+    await this.store?.close()
+  }
+
+  public async addUserProfile(userProfile: UserProfile) {
+    logger('Adding user profile')
+    try {
+      if (!UserProfileStore.validateUserProfile(userProfile)) {
+        // TODO: Send validation errors to frontend or replicate
+        // validation on frontend?
+        logger.error('Failed to add user profile')
+        return
+      }
+      await this.store.put(userProfile.pubKey, userProfile)
+    } catch (err) {
+      logger.error('Failed to add user profile', err)
+    }
+  }
+
+  public static async validateUserProfile(userProfile: UserProfile) {
+    // FIXME: Add additional validation to verify userProfile contains
```

Comment:
I can add a card for it. Basically, I don't know if there is anything enforcing `userProfile` to be a `UserProfile`. I haven't seen anything that would. I think typescript only works at the language level, so it wouldn't let you write a function that references another function if the types were incompatible, but on boundaries (like the network layer) Typescript isn't checking an object at runtime when it comes over the wire. I don't think it's a big deal, because I think the most important validation (that I've thought of) is here, and this function should throw an exception if a field doesn't exist or there is a type mismatch. It's just if we perform object validation, we can provide a more detailed message and catch errors earlier and that's why I've skipped it for now. We can always do better, but I think this is a good start.

Reply:
https://github.com/TryQuiet/quiet/issues/2044